src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2StreamWindowManager.cs

```diff
@@ -0,0 +1,257 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http
+{
+    internal sealed partial class Http2Connection
+    {
+        private struct Http2StreamWindowManager
+        {
+            public const int StreamWindowUpdateRatio = 8;
+            private static readonly double StopWatchToTimesSpan = TimeSpan.TicksPerSecond / (double)Stopwatch.Frequency;
+
+            private readonly Http2Connection _connection;
+            private readonly Http2Stream _stream;
+
+            private readonly double _windowScaleThresholdMultiplier;
+            private readonly int _maxStreamWindowSize;
+            private bool _windowScalingEnabled;
+
+            private int _deliveredBytes;
+            private int _streamWindowSize;
+            private long _lastWindowUpdate;
```

Comment:
You could potentially use `ValueStopwatch` here. It remembers `_startTimestamp` and has `public TimeSpan GetElapsedTime()`. I haven't used it extensively but @MihaZupan did for telemetry/counters.

Reply:
I stole the the unit conversion logic from `ValueStopWatch` because it seemed to be too heavy 😆 We don't need [`_startTimestamp`](https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/Common/src/Extensions/ValueStopwatch/ValueStopwatch.cs#L15) here, we only need differences compared to the last timestamp.

Reply:
A bit hesitant to change & rebenchmark at this point, may try on Monday.

Reply:
What I mean is that:

```C#
TimeSpan dt = _stopwatch.GetElapsedTime();
_stopwatch = ValueStopwatch.StartNew();
```

will call `Stopwatch.GetTimestamp()` (OS call) twice. Although this wouldn't be a noticeable perf issue, I was actually afraid that someone will complain about this in the code review out of perfectionism, so went with manual timestamping instead of `ValueStopwatch` 😄 Don't want to change this now, want to fix my test and the remaining nits, and get this merged ASAP.